lib/Controller/SAMLController.php

```diff
@@ -316,23 +316,37 @@ public function assertionConsumerService() {
 	 * @throws Error
 	 */
 	public function singleLogoutService() {
-
-		$pass = $this->request->passesCSRFCheck();
-		$isGlobalScaleEnabled = $this->config->getSystemValue('gs.enabled', false);
-		$gssMode = $this->config->getSystemValue('gss.mode', '');
-		if (!$pass && $isGlobalScaleEnabled && $gssMode === 'master') {
+		$targetUrl = null;
+		$isFromGS = ($this->config->getSystemValue('gs.enabled', false) &&
+				     $this->config->getSystemValue('gss.mode', '') === 'master');
+		$isFromIDP = !$isFromGS && !empty($_GET['SAMLRequest']);
+
+		if($isFromGS) {
+			// Request is from master GlobalScale
+			// Request validity is check via a JSON Web Token
 			$jwt = $this->request->getParam('jwt', '');
 			$pass = $this->isValidJwt($jwt);
-		}
-
-		if($pass) {
+		} else {
 			$idp = $this->session->get('user_saml.Idp');
 			$auth = new Auth($this->SAMLSettings->getOneLoginSettingsArray($idp));
```

Comment:
Moving this here makes the GS mode fail right @schiessle

Reply:
:open_mouth: really sorry for this newbie mistake. Thanks for your code review, it's fixed in last commits.